File: merge/merge_parallel.jl
#!/usr/bin/env julia
using Base.Threads

# Sequential merge
function merge(left::Vector, right::Vector)
    result = eltype(left)[]
    i, j = 1, 1

    while i <= length(left) && j <= length(right)
        if left[i] <= right[j]
            push!(result, left[i])
            i += 1
        else
            push!(result, right[j])
            j += 1
        end
    end

    append!(result, @view left[i:end])
    append!(result, @view right[j:end])
    return result
end

# Parallel merge sort using threads
function parallel_merge_sort(arr::Vector{T}, depth=0) where T
    n = length(arr)
    if n <= 1
        return arr
    end

    mid = div(n, 2)
    left_half = arr[1:mid]
    right_half = arr[mid+1:end]

    if depth < 3 && n > 1_000
        # Spawn parallel tasks if depth is small and input is large enough
        left_task = Threads.@spawn parallel_merge_sort(left_half, depth + 1)
        right_sorted = parallel_merge_sort(right_half, depth + 1)
        left_sorted = fetch(left_task)
    else
        # Fallback to serial recursion
        left_sorted = parallel_merge_sort(left_half, depth + 1)
        right_sorted = parallel_merge_sort(right_half, depth + 1)
    end

    return merge(left_sorted, right_sorted)
end

arr = [7, 6, 5, 9, 8, 4, 3, 1, 2, 0]
sorted_arr = parallel_merge_sort(arr)
println(sorted_arr)

File: merge/merge_parallel.cpp
#include <iostream>
#include <vector>
#include <thread>
using namespace std;

// Merge function (same as before)
void merge(vector<int>& arr, int left, int mid, int right) {
    int n1 = mid - left + 1;
    int n2 = right - mid;

    vector<int> L(n1), R(n2);
    for (int i = 0; i < n1; i++)
        L[i] = arr[left + i];
    for (int j = 0; j < n2; j++)
        R[j] = arr[mid + 1 + j];

    int i = 0, j = 0, k = left;
    while (i < n1 && j < n2)
        arr[k++] = (L[i] <= R[j]) ? L[i++] : R[j++];

    while (i < n1)
        arr[k++] = L[i++];
    while (j < n2)
        arr[k++] = R[j++];
}

// Parallel Merge Sort
void parallelMergeSort(vector<int>& arr, int left, int right, int depth = 0) {
    if (left >= right)
        return;

    int mid = left + (right - left) / 2;

    // Limit depth to avoid too many threads
    if (depth < 3) {

        thread t1([&arr, left, mid, depth]() {
            parallelMergeSort(arr, left, mid, depth + 1);
        });
        thread t2([&arr, mid, right, depth]() {
            parallelMergeSort(arr, mid + 1, right, depth + 1);
        });
        t1.join();
        t2.join();
    } else {
        // Fallback to serial sort at deeper levels
        parallelMergeSort(arr, left, mid, depth + 1);
        parallelMergeSort(arr, mid + 1, right, depth + 1);
    }

    merge(arr, left, mid, right);
}

// Driver code
int main() {
    int raw[] = {38, 27, 43, 10, 82, 5, 7, 11, 55, 31};
    vector<int> arr(raw, raw + sizeof(raw) / sizeof(raw[0]));
    int n = arr.size();

    parallelMergeSort(arr, 0, n - 1);

    for (int i = 0; i < arr.size(); i++)
        cout << arr[i] << " ";
    cout << endl;

    return 0;
}
